torch/csrc/api/include/torch/detail/ordered_dict.h

```diff
@@ -0,0 +1,204 @@
+#pragma once
+
+#include <ATen/Error.h>
+
+#include <cstdint>
+#include <functional>
+#include <initializer_list>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace detail {
+
+/// A simple ordered dictionary implementation, akin to Python's `OrderedDict`.
+template <typename T>
+class OrderedDict {
+ public:
+  struct Item {
+    Item(std::string key_, T value_)
+        : key(std::move(key_)), value(std::move(value_)) {}
+
+    T& operator*() {
+      return value;
+    }
+    const T& operator*() const {
+      return value;
+    }
+    T* operator->() {
+      return &value;
+    }
+    const T* operator->() const {
+      return &value;
+    }
+
+    const std::string key;
```

Comment:
I... don't particularly mind, but don't I remember you telling me that we should have getters and setters? :^) (Item is a public struct??)

Reply:
That's not a bad idea